backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/VeloxConversion.scala

```diff
@@ -0,0 +1,99 @@
+import org.apache.spark.sql.execution.command.InsertIntoDataSourceDirCommand
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
+import org.apache.spark.sql.execution.datasources.velox.VeloxParquetFileFormat
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.internal.SQLConf.LegacyBehaviorPolicy
+import org.apache.spark.sql.types.{DataType, DateType, Metadata, StructType, TimestampType}
+
+/**
+ * This rule converts vanilla Spark datasource write to velox write, including:
+ * [[InsertIntoHadoopFsRelationCommand]], [[InsertIntoDataSourceDirCommand]]
+ *
+ * Note that, we do not support write partitioned table and bucketed table using velox. More details
+ * see [[GlutenColumnarRules]].
+ */
+case class VeloxConversion(session: SparkSession) extends Rule[LogicalPlan] {
+  private val emptyMetadataHashCode = Metadata.empty.hashCode()
+
+  private def support(options: Map[String, String], schema: StructType): Boolean = {
+    supportConf(options) && supportSchema(schema)
+  }
+
+  private def supportConf(options: Map[String, String]): Boolean = {
+    !options.contains("maxRecordsPerFile") && conf.maxRecordsPerFile == 0
+  }
+
+  private def supportDataType(dt: DataType): Boolean = {
+    val supported = dt match {
+      case _: TimestampType =>
+        // The default behavior of Spark is write INT96 to parquet.
+        // It seems velox does not support read INT64 with TIMESTAMP_MICROS time unit.
+        // todo, enable this after velox support write timestamp as int96
+        // conf.getConf(SQLConf.PARQUET_OUTPUT_TIMESTAMP_TYPE) == "INT96" &&
+        // conf.getConf(SQLConf.PARQUET_INT96_REBASE_MODE_IN_WRITE) !=
+        //   LegacyBehaviorPolicy.LEGACY.toString &&
+        // conf.getConf(SQLConf.PARQUET_REBASE_MODE_IN_WRITE) !=
+        //   LegacyBehaviorPolicy.LEGACY.toString
+        false
+      case _: DateType =>
+        conf.getConf(SQLConf.PARQUET_REBASE_MODE_IN_WRITE) != LegacyBehaviorPolicy.LEGACY.toString
+      case _ => true
+    }
+    supported && new Validator().doSchemaValidate(dt)
+  }
+
+  private def supportSchema(schema: StructType): Boolean = {
+    schema.fields.forall {
+      field =>
+        supportDataType(field.dataType) &&
+        // TODO, do not support write metadata to parquet
+        field.metadata.hashCode() == emptyMetadataHashCode
+    }
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (
+      !conf.getConf(GlutenConfig.GLUTEN_ENABLED) ||
+      !conf.getConf(GlutenConfig.COLUMNAR_PARQUET_WRITE_ENABLED)
+    ) {
+      return plan
+    }
+
+    plan.resolveOperators {
+      case i: InsertIntoHadoopFsRelationCommand
+          if i.fileFormat.isInstanceOf[ParquetFileFormat] &&
+            i.bucketSpec.isEmpty && i.partitionColumns.isEmpty &&
+            support(i.options, StructType.fromAttributes(i.outputColumns)) =>
+        i.copy(fileFormat = new VeloxParquetFileFormat())
```

Comment:
This change will offload all parquet writer job and no need by `.format("velox")` to enable VeloxParquetFileFormat, right?

Reply:
yes, and that's why so many tests failed. After this pr, I think all vanilla Spark parquet write (no partitioned and no bucketed) related tests are covered.

Reply:
https://github.com/oap-project/gluten/pull/2205 is working on attempting  to unify the VeloxParquetFileFormat and HiveFileFormat  two implementations. And put the ParquetFileFormat into the shim layer.  We can enable the config to true in this PR after https://github.com/oap-project/gluten/pull/2205 merged.  Can you also help to review https://github.com/oap-project/gluten/pull/2205?